test/common/grpc/grpc_client_integration_test.cc

```diff
@@ -293,6 +293,15 @@ TEST_P(GrpcClientIntegrationTest, ResourceExhaustedError) {
   dispatcher_helper_.runDispatcher();
 }
 
+// Validate that a trailers UNAUTHENTICATED reply is handled.
+TEST_P(GrpcClientIntegrationTest, UnauthenticatedError) {
+  initialize();
+  auto stream = createStream(empty_metadata_);
+  stream->sendServerInitialMetadata(empty_metadata_);
+  stream->sendServerTrailers(Status::GrpcStatus::Unauthenticated, "error message", empty_metadata_);
```

Comment:
Wouldn't this have worked before if there are no changes to `Status::GrpcStatus`? Can you add a test that uses some fake error code that we don't know about yet to make sure we don't crash? I would expect that test to crash without your fix?

/wait

Reply:
Without my change, the new test would fail because the following expectation would see grpc_status set to -1:
https://github.com/envoyproxy/envoy/blob/97fa8855d115aed81ebe423135b8f023f41f64a7/test/common/grpc/grpc_client_integration_test_harness.h#L144

Which was the cause to the undefined behavior in envoy:
https://github.com/envoyproxy/envoy/blob/97fa8855d115aed81ebe423135b8f023f41f64a7/source/common/grpc/google_async_client_impl.cc#L183-L186

As using -1 as  the index to access an array doesn't always crash envoy, I am not sure if we can have a reliable test to always crash envoy.

Reply:
OK makes sense. My point is can you add a test that returns a response code we don't know about yet, effectively `MaximumValid + 1`. This will test for when we don't update status.h but there is a new code.

Reply:
I take it back. I don't think it's possible to write a test to guard a out-of-sync issue in the future.

The reason is both proto enum (grpc core status code) and c++ enum (envoy grpc status code) don't support programmatically get the current max possible value. Given this restriction, I don't think we can write a test to protect the issue. 

The change I made prevent envoy from crashing in the future even grpc core introduces a new code but envoy status doesn't catch up. I added a new test to cover the case of MaximumValid+1 to make sure envoy won't crash, but it doesn't guard the case when a new grpc code is added but envoy status.h is not updated.

Reply:
The test you added is the one I was looking for. It's definitely not possible to test for out-of-sync, but I want to make sure Envoy doesn't crash if someone adds a new code that Envoy doesn't know about. Your new test tests that. Thank you!